Building burn examples individually breaks

**Describe the bug**
Cd-ing into a mnist or other example crate, and trying to build will result in errors due to incorrect feature flag settings.


**To Reproduce**
1. cd to examples/text-classification
2. cargo build

**Screenshots**
```
   Compiling text-classification v0.10.0 (/Users/dilshod/Projects/burn/examples/text-classification)
error[E0433]: failed to resolve: could not find `train` in `burn`
  --> examples/text-classification/src/training.rs:21:5
   |
21 |     train::{
   |     ^^^^^ could not find `train` in `burn`

error[E0432]: unresolved import `burn::train`
  --> examples/text-classification/src/model.rs:17:5
   |
17 |     train::{ClassificationOutput, TrainOutput, TrainStep, ValidStep},
   |     ^^^^^ could not find `train` in `burn`

error[E0432]: unresolved import `burn::train`
  --> examples/text-classification/src/training.rs:21:5
   |
21 |     train::{
   |     ^^^^^ could not find `train` in `burn`

Some errors have detailed explanations: E0432, E0433.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `text-classification` (lib) due to 3 previous errors

```

**Additional context**

Strangely it works if the entire project is build from the root.

for the "error[E0433]: failed to resolve: could not find `train` in `burn`", usually should add "train" feature in cargo.toml.
but, for your purpose, you should read the examples/text-classification/README.md

> @antimora Thanks for this, I wasn't able to check compilation flags with `cargo check`. Weirdly it always worked when within the workspace, but I'm happy to know we can still validate examples this way.

I think @dae has added examples checks but by default they're not run. If you run example checks, then you can see the failure:

```
[burn]$ ./run-checks.sh examples
    Finished dev [unoptimized + debuginfo] target(s) in 0.23s
     Running `target/debug/xtask run-checks examples`
Checking examples compile 


Checking examples/mnist 


    Checking burn-common v0.10.0 (/Users/dilshod/Projects/burn/burn-common)
    Checking burn-autodiff v0.10.0 (/Users/dilshod/Projects/burn/burn-autodiff)
    Checking burn-core v0.10.0 (/Users/dilshod/Projects/burn/burn-core)
    Checking burn v0.10.0 (/Users/dilshod/Projects/burn/burn)
    Checking mnist v0.10.0 (/Users/dilshod/Projects/burn/examples/mnist)
error[E0433]: failed to resolve: could not find `train` in `burn`
  --> examples/mnist/src/training.rs:12:5
   |
12 |     train::{
   |     ^^^^^ could not find `train` in `burn`

error[E0432]: unresolved import `burn::train`
  --> examples/mnist/src/model.rs:10:5
   |
10 |     train::{ClassificationOutput, TrainOutput, TrainStep, ValidStep},
   |     ^^^^^ could not find `train` in `burn`

error[E0432]: unresolved import `burn::train`
  --> examples/mnist/src/training.rs:12:5
   |
12 |     train::{
   |     ^^^^^ could not find `train` in `burn`

Some errors have detailed explanations: E0432, E0433.
For more information about an error, try `rustc --explain E0432`.
error: could not compile `mnist` (lib) due to 3 previous errors
[burn]$ 

```

--- a/burn-core/Cargo.toml
+++ b/burn-core/Cargo.toml
@@ -33,12 +33,13 @@ dataset-sqlite-bundled = ["burn-dataset/sqlite-bundled"]
 # Backend
 autodiff = ["burn-autodiff"]
 
-ndarray = ["burn-ndarray/default"]
-ndarray-no-std = ["burn-ndarray"]
-ndarray-blas-accelerate = ["ndarray", "burn-ndarray/blas-accelerate"]
-ndarray-blas-netlib = ["ndarray", "burn-ndarray/blas-netlib"]
-ndarray-blas-openblas = ["ndarray", "burn-ndarray/blas-openblas"]
-ndarray-blas-openblas-system = ["ndarray", "burn-ndarray/blas-openblas-system"]
+ndarray = ["__ndarray", "burn-ndarray/default"]
+ndarray-no-std = ["__ndarray", "burn-ndarray"]
+ndarray-blas-accelerate = ["__ndarray", "ndarray", "burn-ndarray/blas-accelerate"]
+ndarray-blas-netlib = ["__ndarray", "ndarray", "burn-ndarray/blas-netlib"]
+ndarray-blas-openblas = ["__ndarray", "ndarray", "burn-ndarray/blas-openblas"]
+ndarray-blas-openblas-system = ["__ndarray", "ndarray", "burn-ndarray/blas-openblas-system"]
+__ndarray = [] # Internal flag to know when one ndarray feature is enabled.
 
 wgpu = ["burn-wgpu"]
 wgpu-autotune = ["wgpu", "burn-wgpu/autotune"]

--- a/burn-core/src/backend.rs
+++ b/burn-core/src/backend.rs
@@ -1,30 +1,60 @@
+#[cfg(any(
+    feature = "ndarray",
+    feature = "ndarray-no-std",
+    feature = "ndarray-blas-accelerate",
+    feature = "ndarray-blas-netlib",
+    feature = "ndarray-blas-openblas",
+    feature = "ndarray-blas-openblas-system",
+))]
 /// Ndarray module.
-#[cfg(feature = "ndarray")]
 pub use burn_ndarray as ndarray;
-#[cfg(feature = "ndarray")]
+
+#[cfg(any(
+    feature = "ndarray",
+    feature = "ndarray-no-std",
+    feature = "ndarray-blas-accelerate",
+    feature = "ndarray-blas-netlib",
+    feature = "ndarray-blas-openblas",
+    feature = "ndarray-blas-openblas-system",
+))]
 /// An NdArrayBackend with a default type of f32.
 pub type NdArrayBackend<F = f32> = ndarray::NdArrayBackend<F>;
+
+#[cfg(all(
+    any(
+        feature = "ndarray",
+        feature = "ndarray-no-std",
+        feature = "ndarray-blas-accelerate",
+        feature = "ndarray-blas-netlib",
+        feature = "ndarray-blas-openblas",
+        feature = "ndarray-blas-openblas-system",
+    ),
+    feature = "autodiff"
+))]
 /// An NdArrayBackend with autodiffing enabled.
-#[cfg(all(feature = "ndarray", feature = "autodiff"))]
 pub type NdArrayAutodiffBackend<F = f32> = crate::autodiff::ADBackendDecorator<NdArrayBackend<F>>;
 
-/// WGPU module.
 #[cfg(feature = "wgpu")]
+/// WGPU module.
 pub use burn_wgpu as wgpu;
+
 #[cfg(feature = "wgpu")]
 /// A WGpuBackend with a default type of f32/i32, and auto graphics.
 pub type WgpuBackend<G = wgpu::AutoGraphicsApi, F = f32, I = i32> = wgpu::WgpuBackend<G, F, I>;
-/// A WgpuBackend with autodiffing enabled.
+
 #[cfg(all(feature = "wgpu", feature = "autodiff"))]
+/// A WgpuBackend with autodiffing enabled.
 pub type WgpuAutodiffBackend<G = wgpu::AutoGraphicsApi, F = f32, I = i32> =
     crate::autodiff::ADBackendDecorator<WgpuBackend<G, F, I>>;
 
+#[cfg(feature = "tch")]
 /// Tch module.
-#[cfg(feature = "tch")]
 pub use burn_tch as tch;
-/// A TchBackend with a default type of f32.
+
 #[cfg(feature = "tch")]
+/// A TchBackend with a default type of f32.
 pub type TchBackend<F = f32> = tch::TchBackend<F>;
-/// A TchBackend with autodiffing enabled.
+
 #[cfg(all(feature = "tch", feature = "autodiff"))]
+/// A TchBackend with autodiffing enabled.
 pub type TchAutodiffBackend<F = f32> = crate::autodiff::ADBackendDecorator<TchBackend<F>>;

--- a/burn-core/src/backend.rs
+++ b/burn-core/src/backend.rs
@@ -1,36 +1,12 @@
-#[cfg(any(
-    feature = "ndarray",
-    feature = "ndarray-no-std",
-    feature = "ndarray-blas-accelerate",
-    feature = "ndarray-blas-netlib",
-    feature = "ndarray-blas-openblas",
-    feature = "ndarray-blas-openblas-system",
-))]
+#[cfg(feature = "__ndarray")]
 /// Ndarray module.
 pub use burn_ndarray as ndarray;
 
-#[cfg(any(
-    feature = "ndarray",
-    feature = "ndarray-no-std",
-    feature = "ndarray-blas-accelerate",
-    feature = "ndarray-blas-netlib",
-    feature = "ndarray-blas-openblas",
-    feature = "ndarray-blas-openblas-system",
-))]
+#[cfg(feature = "__ndarray")]
 /// An NdArrayBackend with a default type of f32.
 pub type NdArrayBackend<F = f32> = ndarray::NdArrayBackend<F>;
 
-#[cfg(all(
-    any(
-        feature = "ndarray",
-        feature = "ndarray-no-std",
-        feature = "ndarray-blas-accelerate",
-        feature = "ndarray-blas-netlib",
-        feature = "ndarray-blas-openblas",
-        feature = "ndarray-blas-openblas-system",
-    ),
-    feature = "autodiff"
-))]
+#[cfg(all(feature = "__ndarray", feature = "autodiff"))]
 /// An NdArrayBackend with autodiffing enabled.
 pub type NdArrayAutodiffBackend<F = f32> = crate::autodiff::ADBackendDecorator<NdArrayBackend<F>>;
 

--- a/examples/mnist/Cargo.toml
+++ b/examples/mnist/Cargo.toml
@@ -17,7 +17,7 @@ tch-gpu = ["burn/tch"]
 wgpu = ["burn/wgpu"]
 
 [dependencies]
-burn = {path = "../../burn", features=["autodiff"]}
+burn = {path = "../../burn", features=["train"]}
 
 # Serialization
 log = {workspace = true}

--- a/examples/text-classification/Cargo.toml
+++ b/examples/text-classification/Cargo.toml
@@ -19,7 +19,7 @@ wgpu = ["burn/wgpu"]
 
 [dependencies]
 # Burn
-burn = {path = "../../burn", features=["autodiff", "ndarray"]}
+burn = {path = "../../burn", features=["train", "ndarray"]}
 
 # Tokenizer
 tokenizers = { version = "0.13.4", default-features = false, features = [

--- a/examples/text-generation/Cargo.toml
+++ b/examples/text-generation/Cargo.toml
@@ -12,7 +12,7 @@ f16 = []
 
 [dependencies]
 # Burn
-burn = {path = "../../burn", features=["autodiff", "tch"]}
+burn = {path = "../../burn", features=["train", "tch"]}
 
 # Tokenizer
 tokenizers = {version = "0.13.4", default-features = false, features = [

